WordPress/gutenberg
# Allow plugins to add new items to the "Add new" menu on the Site Editor

## What?
This PR adds the new `extraTemplateTypes` custom setting and the `extra_template_types` filter which will allow plugins to add new items to the `Add new` dropdown menu on the `Site Editor`.

## Why?
Currently, other plugins cannot add new menu items to the `Add new` menu.

For example, via the WooCommerce plugin, we want to make a block template available in the `Add menu` user interface that will apply to all product attributes pages, which wouldn't be possible right now.

## How?
This PRs is introducing a new filter `extra_template_types` which plugins can use to add their custom templates and that is passed to the Site Editor via a new custom setting called `extraTemplateTypes`.

## Testing Instructions
1. Add a new template using the `extra_template_types` hook:
```php
public function add_extra_template_types( $extra_template_types ) {
	$extra_template_types[] = array(
		'slug'        => 'my_new_template',
		'title'       => 'My new template',
		'description' => 'Displays my new template',
	);

	return $extra_template_types;
}
add_filter( 'extra_template_types', array( $this, 'add_extra_template_types' ) );
```
3. Go to the `Site Editor` > `Browse all templates` (`/wp-admin/site-editor.php?postType=wp_template`)
4. Click on the `Add new` button and make sure you see an item in the dropdown for the `My new template` template.

## Screenshots
| Before | After |
|------|------|
| <img width="327" alt="before" src="https://user-images.githubusercontent.com/186112/205266186-2f925fa1-192f-46dd-b43e-7a9b22291111.png"> | <img width="347" alt="after" src="https://user-images.githubusercontent.com/186112/205266384-1fa838ce-7583-412d-bc58-4c9d15bc44a5.png"> |
